src/Framework/IMetadataContainer.cs

```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Build.Framework
+{
+    /// <summary>
+    /// Provides a way to efficiently enumerate item metadata
+    /// </summary>
+    internal interface IMetadataContainer
+    {
+        /// <summary>
+        /// Returns a list of metadata names and unescaped values, including
+        /// metadata from item definition groups, but not including built-in
+        /// metadata. Implementations should be low-overhead as the method
+        /// is used for serialization (in node packet translator) as well as
+        /// in the binary logger.
+        /// </summary>
+        IEnumerable<KeyValuePair<string, string>> Metadata { get; }
```

Comment:
Naming nit: There's already a [IItem<IMetadatum>.Metadata](Naming nit: [IItem<IMetadatym>]). ProjectItem and ProjectItemInstance both implement it, so by having them implement IMetadataContainer as well they'll end up with two members called `Metadata`, which sounds confusing. Suggestions: `MetadataValues` `RawMetadata`

Reply:
Yes, I had to implement Metadata explicitly on ProjectItemInstance. I think when ProjectItem gets the implementation we can think about renaming. Also it's possible we'll need to rename anyway when Rainer gets back ;)

Having some doubts about MetadataValues (what about names?). RawMetadata on the interface looks a little bit like the interface is a hack to add stuff around something else (and it kind of is...)

We could name it CustomMetadata though. Or convert to a method EnumerateMetadata() or GetCustomMetadata().

Let's see what others think.

Reply:
I like `EnumerateMetadata`. CustomMetadata might throw the user off thinking there's another type of metadata besides direct, item definition, and built-in.